WordPress/gutenberg
# Data: Add redux-routine package for synchronous generator flow

Alternative to #7546, #6999

*WIP:* This pull request is nearly complete, but tests in `packages/data` need to be updated to account for the new options.

This pull request seeks to introduce a new `@wordpress/redux-routine` package, used to create a generator-based coroutine middleware for Redux. The example in the [included README.md](https://github.com/WordPress/gutenberg/blob/40459057a0de958dcf36c4af31b192a0f73ff9e0/packages/redux-routine/README.md) should demonstrate the intent:

```js
import { combineReducers, createStore, applyMiddleware } from 'redux';
import createRoutineMiddleware from '@wordpress/redux-routine';

const middleware = createRoutineMiddleware( {
	async FETCH_JSON( action ) {
		const response = await window.fetch( action.url );
		return response.json();
	},
} );

function temperature( state = null, action ) {
	switch ( action.type ) {
		case 'SET_TEMPERATURE':
			return action.temperature;
	}

	return state;
}

const reducer = combineReducers( { temperature } );

const store = createStore( reducer, applyMiddleware( middleware ) );

function* retrieveTemperature() {
	const result = yield { type: 'FETCH_JSON', url: 'https://' };
	return { type: 'SET_TEMPERATURE', temperature: result.value };
}

store.dispatch( retrieveTemperature() );
```

Atop this, it introduces a new `controls` option to the `registerStore` API which automates the creation of this middleware for stores created using `@wordpress/data`. In the process, it also exposes enhancer introduction (#7417), though I have intentionally chosen to not (yet) document this.

These controls can be used by either `actions` or `resolvers`. I've adapted a few existing `resolvers` in `@wordpress/core-data` as a proof-of-concept.

**Testing instructions:**

Verify there are no regressions in the request of resolved core data. Notably, this includes theme supports and the authors dropdown.

Ensure unit tests pass:

```
npm test
```
